Modules/posixmodule.c

```diff
@@ -5088,212 +5088,235 @@ os__path_splitroot_impl(PyObject *module, path_t *path)
 }
 
 
+#define PY_IFREG 1
+#define PY_IFDIR 2
+#define PY_IFLNK 3
+#define PY_IFMNT 4
+
+static BOOL
+_testFileTypeByHandle(HANDLE hfile, int refFileType, BOOL diskOnly)
+{
+    assert(refFileType == PY_IFREG || refFileType == PY_IFDIR ||
+           refFileType == PY_IFLNK || refFileType == PY_IFMNT);
+
+    DWORD fileDevType = GetFileType(hfile);
+    if ((fileDevType == FILE_TYPE_CHAR) ||
+        (fileDevType != FILE_TYPE_DISK && diskOnly) ||
+        (fileDevType == FILE_TYPE_UNKNOWN && GetLastError() != 0))
+    {
+        return FALSE;
+    }
+
+    FILE_ATTRIBUTE_TAG_INFO info;
+    if (!GetFileInformationByHandleEx(hfile, FileAttributeTagInfo, &info,
+                                      sizeof(info)))
+    {
+        return FALSE;
```

Comment:
Can this fail for regular files?

Reply:
Getting `FileAttributeTagInfo` fails for filesystems that don't support reparse points, such as FAT32, and for the named-pipe and mailslot filesystems (e.g. the root directories "\\\\.\\PIPE\\" and "\\\\.\\MAILSLOT\\"). That's why my example code had the fallback to get `FileBasicInfo`, which is supported by all filesystems. That said, this could be optimized to only try to get `FileAttributeTagInfo` when checking for a symlink or junction, and get `FileBasicInfo` if the latter fails or when checking for a regular file or directory.

Reply:
> That said, this could be optimized to only try to get `FileAttributeTagInfo` when checking for a symlink or junction, and get `FileBasicInfo` if the latter fails

If the latter fails, your old code always returned `False`. So, I simply don't try again with `FileBasicInfo`.
If that wasn't the intention, this needs to be updated.